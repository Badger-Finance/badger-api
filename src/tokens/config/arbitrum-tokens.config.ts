import { Network } from '@badger-dao/sdk';
import { TOKENS } from '../../config/tokens.config';
import { TokenType } from '../enums/token-type.enum';
import { TokenConfig } from '../interfaces/token-config.interface';

export const arbitrumTokensConfig: TokenConfig = {
  [TOKENS.ARB_USDT]: {
    address: TOKENS.ARB_USDT,
    decimals: 6,
    lookupName: 'tether',
    name: 'Tether',
    symbol: 'USDT',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_IBBTC]: {
    address: TOKENS.ARB_IBBTC,
    decimals: 18,
    lookupName: 'interest-bearing-bitcoin',
    name: 'ibBTC',
    symbol: 'ibBTC',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_BADGER]: {
    address: TOKENS.ARB_BADGER,
    decimals: 18,
    lookupName: 'badger-dao',
    name: 'Badger',
    symbol: 'BADGER',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_CRV]: {
    address: TOKENS.ARB_CRV,
    decimals: 18,
    lookupName: 'curve-dao-token',
    name: 'Curve DAO Token',
    symbol: 'CRV',
    type: TokenType.Contract,
  },
  [TOKENS.MULTI_RENBTC]: {
    address: TOKENS.MULTI_RENBTC,
    decimals: 8,
    lookupName: 'renbtc',
    name: 'Ren Protocol BTC',
    symbol: 'renBTC',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_WETH]: {
    address: TOKENS.ARB_WETH,
    decimals: 18,
    lookupName: 'ethereum',
    name: 'Wrapped Ethereum',
    symbol: 'WETH',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_WBTC]: {
    address: TOKENS.ARB_WBTC,
    decimals: 8,
    lookupName: 'bitcoin',
    name: 'Wrapped Bitcoin',
    symbol: 'WBTC',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_SUSHI]: {
    address: TOKENS.ARB_SUSHI,
    decimals: 18,
    lookupName: 'sushi',
    name: 'Sushi',
    symbol: 'SUSHI',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_SWAPR]: {
    address: TOKENS.ARB_SWAPR,
    decimals: 18,
    lookupName: TOKENS.ARB_SWP_SWAPR_WETH,
    name: 'Swapr',
    symbol: 'SWAPR',
    type: TokenType.Contract,
  },
  [TOKENS.ARB_SUSHI_WETH_SUSHI]: {
    address: TOKENS.ARB_SUSHI_WETH_SUSHI,
    decimals: 18,
    lpToken: true,
    name: 'Sushiswap: WETH-SUSHI',
    symbol: 'SLP-WETH-SUSHI',
    type: TokenType.SushiswapLp,
  },
  [TOKENS.ARB_SUSHI_WETH_WBTC]: {
    address: TOKENS.ARB_SUSHI_WETH_WBTC,
    decimals: 18,
    lpToken: true,
    name: 'Sushiswap: WETH-WBTC',
    symbol: 'SLP-WETH-WBTC',
    type: TokenType.SushiswapLp,
  },
  [TOKENS.ARB_CRV_RENBTC]: {
    address: TOKENS.ARB_CRV_RENBTC,
    decimals: 18,
    name: 'Curve.fi: renBTC/wBTC',
    symbol: 'crvrenWBTC',
    type: TokenType.CurveLP,
  },
  [TOKENS.ARB_CRV_TRICRYPTO]: {
    address: TOKENS.ARB_CRV_TRICRYPTO,
    decimals: 18,
    name: 'Curve.fi Tricrypto',
    symbol: 'crvTricrypto',
    type: TokenType.CurveLP,
  },
  [TOKENS.ARB_SWP_SWPR_WETH]: {
    address: TOKENS.ARB_SWP_SWPR_WETH,
    decimals: 18,
    lpToken: true,
    name: 'Swapr SWPR-WETH',
    symbol: 'SWLP-SWPR-WETH',
    type: TokenType.SwaprLp,
  },
  [TOKENS.ARB_SWP_WBTC_WETH]: {
    address: TOKENS.ARB_SWP_WBTC_WETH,
    decimals: 18,
    lpToken: true,
    name: 'Swapr WBTC-WETH',
    symbol: 'SWLP-WBTC-WETH',
    type: TokenType.SwaprLp,
  },
  [TOKENS.ARB_SWP_BADGER_WETH]: {
    address: TOKENS.ARB_SWP_BADGER_WETH,
    decimals: 18,
    lpToken: true,
    name: 'Swapr BADGER-WETH',
    symbol: 'SWLP-BADGER-WETH',
    type: TokenType.SwaprLp,
  },
  [TOKENS.ARB_SWP_IBBTC_WETH]: {
    address: TOKENS.ARB_SWP_IBBTC_WETH,
    decimals: 18,
    lpToken: true,
    name: 'Swapr ibBTC-WETH',
    symbol: 'SWLP-IBBTC-WETH',
    type: TokenType.SwaprLp,
  },
  [TOKENS.BARB_SUSHI_WETH_SUSHI]: {
    address: TOKENS.BARB_SUSHI_WETH_SUSHI,
    decimals: 18,
    name: 'bSushiswap: WETH-SUSHI',
    symbol: 'bSLP-WETH-SUSHI',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_SUSHI_WETH_SUSHI,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_SUSHI_WETH_WBTC]: {
    address: TOKENS.BARB_SUSHI_WETH_WBTC,
    decimals: 18,
    name: 'bSushiswap: WETH-WBTC',
    symbol: 'bSLP-WETH-WBTC',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_SUSHI_WETH_WBTC,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_CRV_RENBTC]: {
    address: TOKENS.BARB_CRV_RENBTC,
    decimals: 18,
    name: 'bCurve.fi wBTC/renBTC',
    symbol: 'bcrvrenBTC',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_CRV_RENBTC,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_CRV_TRICRYPTO]: {
    address: TOKENS.BARB_CRV_TRICRYPTO,
    decimals: 18,
    name: 'bCurve.fi Tricrypto',
    symbol: 'bcrvTricrypto',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_CRV_TRICRYPTO,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_CRV_TRICRYPTO_LITE]: {
    address: TOKENS.BARB_CRV_TRICRYPTO_LITE,
    decimals: 18,
    name: 'bCurve.fi Tricrypto Light',
    symbol: 'bcrvTricryptoLight',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_CRV_TRICRYPTO,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_SWP_SWPR_WETH]: {
    address: TOKENS.BARB_SWP_SWPR_WETH,
    decimals: 18,
    name: 'bSwapr SWPR-WETH',
    symbol: 'bSWLP-SWPR-WETH',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_SWP_SWAPR_WETH,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_SWP_WBTC_WETH]: {
    address: TOKENS.BARB_SWP_WBTC_WETH,
    decimals: 18,
    name: 'bSwapr WBTC-WETH',
    symbol: 'bSWLP-WBTC-WETH',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_SWP_WBTC_WETH,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_SWP_BADGER_WETH]: {
    address: TOKENS.BARB_SWP_BADGER_WETH,
    decimals: 18,
    name: 'bSwapr BADGER-WETH',
    symbol: 'bSWLP-BADGER-WETH',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_SWP_BADGER_WETH,
      network: Network.Arbitrum,
    },
  },
  [TOKENS.BARB_SWP_IBBTC_WETH]: {
    address: TOKENS.BARB_SWP_IBBTC_WETH,
    decimals: 18,
    name: 'bSwapr ibBTC-WETH',
    symbol: 'bSWLP-IBBTC-WETH',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.ARB_SWP_IBBTC_WETH,
      network: Network.Arbitrum,
    },
  },
};
