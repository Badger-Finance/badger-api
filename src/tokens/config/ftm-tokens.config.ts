import { TokenConfig } from '../interfaces/token-config.interface';

export const fantomTokensConfig: TokenConfig = {};
