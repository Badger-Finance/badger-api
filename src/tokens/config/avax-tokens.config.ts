import { Network } from '@badger-dao/sdk';
import { TOKENS } from '../../config/tokens.config';
import { TokenType } from '../enums/token-type.enum';
import { TokenConfig } from '../interfaces/token-config.interface';

export const avalancheTokensConfig: TokenConfig = {
  [TOKENS.AVAX_WBTC]: {
    address: TOKENS.AVAX_WBTC,
    decimals: 8,
    lookupName: 'wrapped-bitcoin',
    name: 'Wrapped Bitcoin',
    symbol: 'WBTC',
    type: TokenType.Contract,
  },
  [TOKENS.BAVAX_WBTC]: {
    address: TOKENS.BAVAX_WBTC,
    decimals: 18,
    name: 'bWBTC',
    symbol: 'bWTBC',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.AVAX_WBTC,
      network: Network.Avalanche,
    },
  },
};
