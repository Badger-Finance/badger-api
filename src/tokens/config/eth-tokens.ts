import { TOKENS } from '../../config/constants';
import { Token } from '../../interface/Token';
import { getContractPrice, getTokenPrice } from '../../prices/PricesService';
import { getSushiswapPrice, getUniswapPrice } from '../../protocols/common/swap-util';

const badger: Token = {
  address: TOKENS.BADGER,
  name: 'Badger',
  symbol: 'BADGER',
  decimals: 18,
  price: () => getContractPrice(TOKENS.BADGER),
};

const digg: Token = {
  address: TOKENS.DIGG,
  name: 'Digg',
  symbol: 'DIGG',
  decimals: 9,
  price: () => getContractPrice(TOKENS.DIGG),
};

const sushiWbtcDigg: Token = {
  address: TOKENS.SUSHI_DIGG_WBTC,
  name: 'SushiSwap: WBTC-DIGG',
  symbol: 'SushiSwap WBTC/DIGG LP (SLP)',
  decimals: 18,
  price: () => getSushiswapPrice(TOKENS.SUSHI_DIGG_WBTC),
};

const sushiBagderDigg: Token = {
  address: TOKENS.SUSHI_BADGER_WBTC,
  name: 'SushiSwap: WBTC-BADGER',
  symbol: 'Badger Sett SushiSwap LP Token (bSLP)',
  decimals: 18,
  price: () => getSushiswapPrice(TOKENS.SUSHI_BADGER_WBTC),
};

const sushiWbtcEth: Token = {
  address: TOKENS.SUSHI_ETH_WBTC,
  name: 'SushiSwap: WBTC-ETH',
  symbol: 'SushiSwap WBTC/ETH LP (SLP)',
  decimals: 18,
  price: () => getSushiswapPrice(TOKENS.SUSHI_ETH_WBTC),
};

const uniWbtcDigg: Token = {
  address: TOKENS.UNI_DIGG_WBTC,
  name: 'Uniswap V2: WBTC-DIGG',
  symbol: 'Uniswap WBTC/DIGG LP (UNI-V2)',
  decimals: 18,
  price: () => getUniswapPrice(TOKENS.UNI_DIGG_WBTC),
};

const uniBagderDigg: Token = {
  address: TOKENS.UNI_BADGER_WBTC,
  name: 'Uniswap V2: WBTC-BADGER',
  symbol: 'Uniswap WBTC/BADGER LP (UNI-V2)',
  decimals: 18,
  price: () => getUniswapPrice(TOKENS.UNI_BADGER_WBTC),
};

const crvRenBtc: Token = {
  address: TOKENS.CRV_RENBTC,
  name: 'Curve.fi: renCrv Token',
  symbol: 'Curve.fi renBTC/wBTC (crvRenWBTC)',
  decimals: 18,
  price: () => getTokenPrice(TOKENS.WBTC),
};

const crvTbtc: Token = {
  address: TOKENS.CRV_TBTC,
  name: 'Curve.fi tBTC/sbtcCrv',
  symbol: 'Curve.fi tBTC/sbtcCrv (tbtc/sbtc)',
  decimals: 18,
  price: () => getTokenPrice(TOKENS.WBTC),
};

const crvSbtc: Token = {
  address: TOKENS.CRV_SBTC,
  name: 'Curve.fi renBTC/wBTC/sBTC',
  symbol: 'Curve.fi renBTC/wBTC/sBTC',
  decimals: 18,
  price: () => getTokenPrice(TOKENS.WBTC),
};

export const ethTokens = [
  badger,
  digg,
  sushiWbtcDigg,
  sushiBagderDigg,
  sushiWbtcEth,
  uniWbtcDigg,
  uniBagderDigg,
  crvRenBtc,
  crvTbtc,
  crvSbtc,
];
