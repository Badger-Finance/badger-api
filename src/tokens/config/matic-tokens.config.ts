import { ChainNetwork } from '../../chains/enums/chain-network.enum';
import { TOKENS } from '../../config/tokens.config';
import { TokenType } from '../enums/token-type.enum';
import { TokenConfig } from '../interfaces/token-config.interface';

export const maticTokensConfig: TokenConfig = {
  [TOKENS.MATIC_RENBTC]: {
    address: TOKENS.MATIC_RENBTC,
    decimals: 8,
    lookupName: 'renbtc',
    name: 'Ren Protocol BTC',
    symbol: 'renBTC',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_WBTC]: {
    address: TOKENS.MATIC_WBTC,
    decimals: 8,
    lookupName: 'wrapped-bitcoin',
    name: 'Wrapped Bitcoin',
    symbol: 'WBTC',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_USDC]: {
    address: TOKENS.MATIC_USDC,
    decimals: 6,
    lookupName: 'usd-coin',
    name: 'US Dollar Coin',
    symbol: 'USDC',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_USDT]: {
    address: TOKENS.MATIC_USDT,
    decimals: 6,
    lookupName: 'tether',
    name: 'Tether',
    symbol: 'USDT',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_DAI]: {
    address: TOKENS.MATIC_DAI,
    decimals: 18,
    lookupName: 'dai',
    name: 'Dai',
    symbol: 'DAI',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_IBBTC]: {
    address: TOKENS.MATIC_IBBTC,
    decimals: 18,
    lookupName: 'interest-bearing-bitcoin',
    name: 'ibBTC',
    symbol: 'ibBTC',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_AMWBTC]: {
    address: TOKENS.MATIC_AMWBTC,
    decimals: 8,
    lookupName: 'aave-polygon-wbtc',
    name: 'Aave WBTC',
    symbol: 'amWBTC',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_AMWETH]: {
    address: TOKENS.MATIC_AMWETH,
    decimals: 18,
    lookupName: 'aave-polygon-weth',
    name: 'Aave WETH',
    symbol: 'amWETH',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_AMDAI]: {
    address: TOKENS.MATIC_AMDAI,
    decimals: 18,
    lookupName: 'aave-polygon-dai',
    name: 'Aave DAI',
    symbol: 'amDAI',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_AMUSDC]: {
    address: TOKENS.MATIC_AMUSDC,
    decimals: 6,
    lookupName: 'aave-polygon-usdc',
    name: 'Aave USDC',
    symbol: 'amUSDC',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_AMUSDT]: {
    address: TOKENS.MATIC_AMUSDT,
    decimals: 6,
    lookupName: 'aave-polygon-usdt',
    name: 'Aave USDT',
    symbol: 'amUSDT',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_CRV]: {
    address: TOKENS.MATIC_CRV,
    decimals: 18,
    lookupName: 'curve-dao-token',
    name: 'Curve DAO Token',
    symbol: 'CRV',
    type: TokenType.Contract,
  },
  [TOKENS.MATIC_SUSHI_IBBTC_WBTC]: {
    address: TOKENS.MATIC_SUSHI_IBBTC_WBTC,
    decimals: 18,
    lpToken: true,
    name: 'Sushiswap: ibBTC-WBTC',
    symbol: 'SLP-IBBTC-WBTC',
    type: TokenType.SushiswapLp,
  },
  [TOKENS.MATIC_QUICK_USDC_WBTC]: {
    address: TOKENS.MATIC_QUICK_USDC_WBTC,
    decimals: 18,
    lpToken: true,
    name: 'Quickswap: USDC-WBTC',
    symbol: 'QLP-USDC-WBTC',
    type: TokenType.SushiswapLp,
  },
  [TOKENS.MATIC_CRV_AM3CRV]: {
    address: TOKENS.MATIC_CRV_AM3CRV,
    decimals: 18,
    name: 'Curve.fi amDAI/amUSDC/amUSDT',
    symbol: 'crvAM3',
    type: TokenType.CurveLP,
  },
  [TOKENS.MATIC_CRV_AMWBTC]: {
    address: TOKENS.MATIC_CRV_AMWBTC,
    decimals: 18,
    name: 'Curve.fi amWBTC/renBTC',
    symbol: 'btcCRV',
    type: TokenType.CurveLP,
  },
  [TOKENS.BMATIC_SUSHI_IBBTC_WBTC]: {
    address: TOKENS.BMATIC_SUSHI_IBBTC_WBTC,
    decimals: 18,
    name: 'bSushiSwap: ibBTC-WBTC',
    symbol: 'bSLP-IBBTC-WBTC',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.MATIC_SUSHI_IBBTC_WBTC,
      network: ChainNetwork.Matic,
    },
  },
  [TOKENS.BMATIC_QUICK_USDC_WBTC]: {
    address: TOKENS.BMATIC_QUICK_USDC_WBTC,
    decimals: 18,
    name: 'bQuickswap: USDC-WBTC',
    symbol: 'bQLP-USDC-WBTC',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.MATIC_QUICK_USDC_WBTC,
      network: ChainNetwork.Matic,
    },
  },
  [TOKENS.BMATIC_CRV_AMWBTC]: {
    address: TOKENS.BMATIC_CRV_AMWBTC,
    decimals: 18,
    name: 'bCurve.fi amWBTC/renBTC',
    symbol: 'bbtcCRV',
    type: TokenType.Vault,
    vaultToken: {
      address: TOKENS.MATIC_CRV_AMWBTC,
      network: ChainNetwork.Matic,
    },
  },
};
