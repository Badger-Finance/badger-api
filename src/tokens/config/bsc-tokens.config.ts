import { ChainNetwork } from '../../chains/enums/chain-network.enum';
import { TOKENS } from '../../config/constants';
import { TokenType } from '../enums/token-type.enum';
import { TokenConfig } from '../types/token-config.type';
import { ethTokensConfig } from './eth-tokens.config';

export const bscTokensConfig: TokenConfig = {
  [TOKENS.CAKE]: {
    address: TOKENS.CAKE,
    decimals: 18,
    lookupName: 'pancakeswap-token',
    name: 'Cake',
    symbol: 'CAKE',
    type: TokenType.Contract,
  },
  [TOKENS.WBNB]: {
    address: TOKENS.WBNB,
    decimals: 18,
    lookupName: 'binancecoin',
    name: 'Wrapped Binance Coin',
    symbol: 'WBNB',
    type: TokenType.Contract,
  },
  [TOKENS.BTCB]: {
    address: TOKENS.BTCB,
    decimals: 18,
    lookupName: 'binance-bitcoin',
    name: 'Binance Pegged Bitcoin',
    symbol: 'BTCB',
    type: TokenType.Contract,
  },
  [TOKENS.PANCAKE_BNB_BTCB]: {
    address: TOKENS.PANCAKE_BNB_BTCB,
    decimals: 18,
    lpToken: true,
    name: 'Pancakeswap: WBNB-BTCB',
    symbol: 'Pancakeswap WBNB/BTCB LP (PLP)',
    type: TokenType.PancakeswapLp,
  },
  [TOKENS.PANCAKE_BBADGER_BTCB]: {
    address: TOKENS.PANCAKE_BBADGER_BTCB,
    decimals: 18,
    lpToken: true,
    name: 'Pancakeswap: BBADGER-BTCB',
    symbol: 'Pancakeswap BBADGER/BTCB LP (PLP)',
    type: TokenType.PancakeswapLp,
  },
  [TOKENS.PANCAKE_BDIGG_BTCB]: {
    address: TOKENS.PANCAKE_BDIGG_BTCB,
    decimals: 18,
    lpToken: true,
    name: 'Pancakeswap: BDIGG-BTCB',
    symbol: 'Pancakeswap BDIGG/BTCB LP (PLP)',
    type: TokenType.PancakeswapLp,
  },
  [TOKENS.BSC_BBADGER]: {
    address: TOKENS.BSC_BBADGER,
    decimals: 18,
    name: 'bBadger',
    symbol: 'bBADGER',
    type: TokenType.Wrapper,
    vaultToken: {
      symbol: ethTokensConfig[TOKENS.BADGER].symbol,
      network: ChainNetwork.Ethereum,
    },
  },
  [TOKENS.BSC_BDIGG]: {
    address: TOKENS.BSC_BDIGG,
    decimals: 18,
    name: 'bDigg',
    symbol: 'bDIGG',
    type: TokenType.Wrapper,
    vaultToken: {
      symbol: ethTokensConfig[TOKENS.DIGG].symbol,
      network: ChainNetwork.Ethereum,
    },
  },
};
