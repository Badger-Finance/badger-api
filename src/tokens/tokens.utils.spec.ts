import { NotFound } from '@tsed/exceptions';
import { Ethereum } from '../chains/config/eth.config';
import { TOKENS } from '../config/tokens.config';
import * as swapUtils from '../protocols/common/swap.utils';
import { getSettDefinition } from '../setts/setts.utils';
import { ethTokensConfig } from './config/eth-tokens.config';
import { formatBalance, getSettUnderlyingTokens, getToken, getTokenByName } from './tokens.utils';

describe('token.utils', () => {
  const rook = '0xfA5047c9c78B8877af97BDcb85Db743fD7313d4a';

  describe('getToken', () => {
    describe('lookup invalid token address', () => {
      it('throws a not found error', () => {
        expect(() => getToken(rook)).toThrow(NotFound);
        expect(() => getToken(rook)).toThrow(`${rook} not supported`);
      });
    });
    describe('lookup supported token address', () => {
      it('returns the requested token information', () => {
        const expected = ethTokensConfig[TOKENS.BADGER];
        const actual = getToken(expected.address);
        expect(actual).toMatchObject(expected);
      });
    });
  });

  describe('getTokenByName', () => {
    describe('lookup invalid token name', () => {
      it('throws a not found error', () => {
        const token = 'invalid-token';
        expect(() => getTokenByName(token)).toThrow(NotFound);
        expect(() => getTokenByName(token)).toThrow(`${token} not supported`);
      });
    });
    describe('lookup supported token name', () => {
      it('returns the requested token information', () => {
        const expected = getToken(TOKENS.BADGER);
        const actual = getTokenByName(expected.name);
        expect(actual).toMatchObject(expected);
      });
    });
  });

  describe('getSettUnderlyingTokens', () => {
    const chain = new Ethereum();

    describe('lookup invalid lp token sett definition', () => {
      it('throws a not found error', async () => {
        const definition = getSettDefinition(chain, TOKENS.BSUSHI_ETH_WBTC);
        // simulate an error looking up token liquidity
        jest.spyOn(swapUtils, 'getLiquidityData').mockImplementation((_chain, _contract) => {
          throw new Error();
        });
        await expect(getSettUnderlyingTokens(chain, definition)).rejects.toThrow(NotFound);
      });
    });
    describe('lookup single asset sett', () => {
      it('returns 1 underlying token', async () => {
        const definition = getSettDefinition(chain, TOKENS.BBADGER);
        const tokens = await getSettUnderlyingTokens(chain, definition);
        expect(tokens.length).toEqual(1);
        expect(tokens[0]).toMatchObject(ethTokensConfig[TOKENS.BADGER]);
      });
    });
    describe('lookup mutli asset sett', () => {
      it('returns multiple underlying tokens', async () => {
        const definition = getSettDefinition(chain, TOKENS.BSUSHI_ETH_WBTC);
        // simulate looking up token liquidity
        jest.spyOn(swapUtils, 'getLiquidityData').mockImplementation(async (_chain, contract) => ({
          contract,
          token0: TOKENS.WETH,
          token1: TOKENS.WBTC,
          reserve0: 1,
          reserve1: 1,
          totalSupply: 1,
        }));
        const tokens = await getSettUnderlyingTokens(chain, definition);
        expect(tokens.length).toBeGreaterThan(1);
        const expected = [TOKENS.WBTC, TOKENS.WETH].sort();
        expect(tokens.map((t) => t.address).sort()).toMatchObject(expected);
      });
    });
  });

  describe('formatBalance', () => {
    describe('given valid input', () => {
      it.each([
        ['12345000000000000000', undefined, 12.345],
        ['12345000000000000000', 15, 12345],
        ['12345000000000000000', 12, 12345000],
        ['12345000000000000000', 21, 0.012345],
        ['12345000033000000000', 15, 12345.000033],
        ['00000000000000000000', undefined, 0],
        ['00000000000000000000', 15, 0],
        ['00000000000000000000', 21, 0],
      ])('Evaluates %s with %i decimals as %f', (input, decimals, result) => {
        expect(result).toEqual(formatBalance(input, decimals));
      });
    });
  });
});
