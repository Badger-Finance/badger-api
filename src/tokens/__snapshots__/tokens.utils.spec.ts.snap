// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`token.utils getVaultTokens returns all deposit token for a liquidity token underlying token 1`] = `
Array [
  Object {
    "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "balance": 1,
    "decimals": 8,
    "name": "Wrapped Bitcoin",
    "symbol": "WBTC",
    "value": 550,
  },
  Object {
    "address": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    "balance": 20,
    "decimals": 18,
    "name": "Wrapped Ethereum",
    "symbol": "WETH",
    "value": 61480,
  },
]
`;

exports[`token.utils getVaultTokens returns the single underlying token for a non liquidity token underlying token 1`] = `
Array [
  Object {
    "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "balance": 10,
    "decimals": 18,
    "name": "Badger",
    "symbol": "BADGER",
    "value": 8390,
  },
]
`;
